src/metabase/integrations/slack.clj

```diff
@@ -319,26 +319,38 @@
+   the URL of the uploaded file."
   [file       :- NonEmptyByteArray
    filename   :- ms/NonBlankString
    channel-id :- ms/NonBlankString]
   {:pre [(slack-configured?)]}
-  (let [request  {:multipart [{:name "file",     :content file}
-                              {:name "filename", :content filename}
-                              {:name "channels", :content channel-id}]}
-        response (try
-                   (POST "files.upload" request)
-                   (catch Throwable e
-                     ;; If file upload fails with a "not_in_channel" error, we join the channel and try again.
-                     ;; This is expected to happen the first time a Slack subscription is sent.
-                     (if (= "not_in_channel" (:error-code (ex-data e)))
-                       (do (-> channel-id
-                               (maybe-lookup-id (slack-cached-channels-and-usernames))
-                               join-channel!)
-                           (POST "files.upload" request))
-                       (throw e))))]
-    (u/prog1 (get-in response [:file :url_private])
+  ;; TODO: we could make uploading files a lot faster by uploading the files in parallel.
+  ;; Steps 1 and 2 can be done for all files in parallel, and step 3 can be done once at the end.
+  (let [;; Step 1: Get the upload URL using files.getUploadURLExternal
+        {upload-url :upload_url
+         file-id    :file_id} (POST "files.getUploadURLExternal" {:query-params {:filename filename
+                                                                                 :length   (count file)}})
+        ;; Step 2: Upload the file to the obtained upload URL
+        _ (let [{:keys [status body]} (http/post upload-url {:multipart [{:name "file" :content file}]})]
+            (when (not= 200 status)
+              (throw (ex-info "Failed to upload file to Slack:" {:status status, :body body}))))
+        ;; Step 3: Complete the upload using files.completeUploadExternal
+        channel-id (maybe-lookup-id channel-id (slack-cached-channels-and-usernames))
+        complete! (fn []
+                    (POST "files.completeUploadExternal"
+                      {:query-params {:files      (json/generate-string [{:id file-id, :title filename}])
+                                      :channel_id channel-id}}))
+        complete-response (try
+                            (complete!)
+                            (catch Throwable e
+                              ;; If file upload fails with a "not_in_channel" error, we join the channel and try again.
+                              ;; This is expected to happen the first time a Slack subscription is sent.
+                              (if (= "not_in_channel" (:error-code (ex-data e)))
+                                (do (join-channel! channel-id)
+                                    (complete!))
+                                (throw e))))]
+    (u/prog1 (get-in complete-response [:files 0 :url_private])
```

Comment:
I think this is a lot easier to read with named functions and less complex `let` destructuring. Something like:

```clojure
(defn slack-upload-url
  [filename file]
  (POST "files.getUploadExternal" {:query-params {:filename filename
                                                  :length   (count file)}}))

(defn upload-file-or-throw!
  [file {upload-url :upload_url file-id :file_id}]
  (let [response (http/post upload-url {:multipart [{:name "file"
                                                     :content "file"}]})]
    (if (= (:status response) 200)
      response
      (throw (ex-info "Failed to upload file to Slack:" (select-keys response [:status :body]))))))

(defn complete-upload!*
  [channel-id filename file-id]
  (POST "files.completeUploadExternal"
      {:query-params {:files      (json/generate-string [{:id file-id :title filename}])
                      :channel_id channel-id}}))

(defn complete-upload!
  [channel-id filename file-id]
  (try
    (complete-upload!* channel-id filename file-id)
    (catch Throwable e
      ;; If file upload fails with a [etc]
      (if (= (:error-code (ex-data e)) "not_in_channel")
        (doto channel-id
          (join-channel!)
          (complete-upload!* filename file-id))
        (throw e)))))


(defn upload-file!
  [file filename channel-id]
  (let [external-url      (slack-upload-url filename file)
        _                 (upload-file-or-throw! file external-url)
        complete-response (complete-upload! (maybe-lookup-id channel-id (slack-cached-channels-and-usernames))
                                            filename
                                            (:file_id external-url))
        image-url (get-in complete-response [:files 0 :url_private])]
    (log/debug "Uploaded image" image-url)
    image-url))
```

Reply:
I didn't transcribe all the comments / write docstrings / etc (and I should have) but you get the idea

Reply:
Thanks, this is good stuff. I'll try it

Reply:
[f351704](https://github.com/metabase/metabase/pull/41974/commits/f351704195fa0cd4b43be83353921dfb403fc34d)